File: p428h/codec_n_ary_tree.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <vector>
#include <string>
#include <sstream>
#include <iostream>

using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
class Codec {
public:
    // Encodes a tree to a single string.
    string serialize(Node* root) {
        string data = "";
        function<void(Node*)> recHelper = [&](Node* node) -> void {
            if (!node) return;
            data += " " + to_string(node->val)
                  + " " + to_string(node->children.size());
            for (auto child: node->children) {
                recHelper(child);
            }
        };
        recHelper(root);
        return data;
    }

    // Decodes your encoded data to tree.
    Node* deserialize(string data) {
        stringstream ss(data);
        function<Node*(stringstream&)> recHelper = [&](stringstream& ss) -> Node* {
            int val, size;
            if (ss >> val >> size) {
                vector<Node*> children;
                for (int i = 0; i < size; ++i) {
                    children.push_back(recHelper(ss));
                }
                return new Node(val, children);
            } else {
                return nullptr;
            }
        };
        return recHelper(ss);
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));
int main(int argc, char* argv[]) {
    vector<string> tests = {
        "",
        " 1 0",
        " 1 1 2 0",
        " 1 2 2 0 3 0",
        " 1 3 2 0 3 0 4 0",
        " 1 10 2 0 3 0 4 0 5 0 6 0 7 0 8 0 9 0 10 0 11 0",
        " 1 1 2 1 3 0",
        " 1 3 3 2 5 0 6 0 2 0 4 0"
    };
    Codec codec;
    for (auto& t: tests) {
        Node* tree = codec.deserialize(t);
        cout << "Deserialize/Serialize " << t << " -> "
             << boolalpha << (codec.serialize(tree) == t) << endl;
    }
    return 0;
}

File: p428h/codec_n_ary_tree.js
/**
 * // Definition for a Node.
 */
function Node(val, children) {
  this.val = val;
  this.children = children;
}

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
function serialize(root) {
  let data = '';
  // Helper function takes closure of data
  const recHelper = node => {
    if (node == null) return;
    data += ' ' + node.val + ' ' + node.children.length;
    for (let child of node.children) {
      recHelper(child);
    }
  };
  recHelper(root);
  return data.trim();
}

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
function deserialize(data) {
  if (data === undefined || data.length <= 0) return null;
  let buffer = data.split(' ');
  const recHelper = buffer => {
    if (buffer.length === 0) return null;
    if (buffer.length === 1) return new Node(parseInt(buffer[0]), []);
    let [root, size] = buffer;
    buffer.shift();
    buffer.shift();
    let children = [];
    for (let i = 0; i < size; ++i) {
      children.push(recHelper(buffer));
    }
    return new Node(parseInt(root), children);
  };
  return recHelper(buffer);
}

// TEST
[
  '',
  '1 0',
  '1 1 2 0',
  '1 2 2 0 3 0',
  '1 3 2 0 3 0 4 0',
  '1 10 2 0 3 0 4 0 5 0 6 0 7 0 8 0 9 0 10 0 11 0',
  '1 1 2 1 3 0',
  '1 3 3 2 5 0 6 0 2 0 4 0'
].forEach(t => {
  const tree = deserialize(t);
  console.log('Deserialize/Serialize', t, '->', serialize(tree) === t);
});
